geomstats/geometry/pullback_metric.py

```diff
@@ -103,6 +103,415 @@ def pickable_inner_product(i, j):
             pickable_inner_product(i, j)
             for i, j in itertools.product(range(self.dim), range(self.dim))
         )
-
         metric_mat = gs.reshape(gs.array(out), (-1, self.dim, self.dim))
         return metric_mat[0] if base_point.ndim == 1 else metric_mat
+
+
+class PullbackDiffeoMetric(RiemannianMetric, abc.ABC):
+    """
+    Pullback metric via a diffeomorphism.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension.
+    """
+
+    def __init__(self, dim, shape):
+        super(PullbackDiffeoMetric, self).__init__(dim=dim, shape=shape)
+
+        # Inner variable
+        self._embedding_metric = None
+        self._jacobian_diffeomorphism = None
+        self._inverse_jacobian_diffeomorphism = None
+
+        self.shape_dim = int(gs.prod(gs.array(shape)))
+        self.embedding_space_shape_dim = int(
+            gs.prod(gs.array(self.embedding_metric.shape))
+        )
+
+    @abc.abstractmethod
+    def create_embedding_metric(self):
+        r"""Create the metric this metric is in diffeomorphism with.
+
+        This instantiate the metric to use as image space of the
+        diffeomorphism.
+
+        -------
+        embedding_metric : RiemannianMetric object
+            The metric of the embedding space
+        """
+
+    @property
+    def embedding_metric(self):
+        r"""Property wrapper around the metric.
+
+        -------
+        embedding_metric : RiemannianMetric object
+            The metric of the embedding space
+        """
+        if self._embedding_metric is None:
+            self._embedding_metric = self.create_embedding_metric()
+        return self._embedding_metric
+
+    @abc.abstractmethod
+    def diffeomorphism(self, base_point):
+        r"""Diffeomorphism at base point.
+
+        Let :math:`f` be the diffeomorphism
+        :math:`f: M \rightarrow N` of the manifold
+        :math:`M` into the manifold `N`.
+
+        Parameters
+        ----------
+        base_point : array-like, shape=[..., *shape]
+            Base point.
+
+        Returns
+        -------
+        image_point : array-like, shape=[..., *i_shape]
+            Inner-product matrix.
+        """
+
+    def jacobian_diffeomorphism(self, base_point):
+        r"""Jacobian of the diffeomorphism at base point.
+
+        Let :math:`f` be the diffeomorphism
+        :math:`f: M \rightarrow N` of the manifold
+        :math:`M` into the manifold `N`.
+
+        Parameters
+        ----------
+        base_point : array-like, shape=[..., *shape]
+            Base point.
+
+        Returns
+        -------
+        mat : array-like, shape=[..., *i_shape, *shape]
+            Inner-product matrix.
+        """
+        rad = base_point.shape[: -len(self.shape)]
+        base_point = gs.reshape(base_point, (-1,) + self.shape)
+        # Only compute graph once
+        if self._jacobian_diffeomorphism is None:
```

Comment:
Can we do similar to `embedding_metric` with a property?

Reply:
Okay I will just need to find a proper naming to avoid confusion between raw jacobian and the method around it. Also, I made this choice at first because, some inherited classes can overide the method so does not need this autodiff attribute !